File: scripts/download-reader-data.py
#!/usr/bin/env python3
"""Download reader JSON data from GitHub raw CDN for Vercel builds."""
import urllib.request
import json
import os
import sys
import time
from concurrent.futures import ThreadPoolExecutor, as_completed

BASE = "https://raw.githubusercontent.com/petteknanach/ajew-org/main/public/reader"
DEST = "public/reader"

def dl(filepath):
    url = BASE + "/" + filepath
    try:
        req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
        with urllib.request.urlopen(req, timeout=30) as resp:
            if resp.status != 200:
                return None
            data = resp.read()
            dest = os.path.join(DEST, filepath)
            os.makedirs(os.path.dirname(dest), exist_ok=True)
            with open(dest, "wb") as f:
                f.write(data)
            return (filepath, len(data))
    except Exception as e:
        if "404" not in str(e):
            print(f"  ERR {filepath}: {e}", flush=True)
        return None

def get_index(book_part):
    idx_path = book_part + "/index.json"
    result = dl(idx_path)
    if not result:
        return None
    with open(os.path.join(DEST, idx_path)) as f:
        return json.load(f)

def download_book_part(book_part):
    index = get_index(book_part)
    if not index:
        return 0

    files_to_dl = []
    
    # Download intro sections
    for intro in index.get("introSections", []):
        slug = intro.get("slug", intro.get("file"))
        if slug:
            files_to_dl.append(book_part + "/" + slug + ".json")
    
    # Download preface sections
    for preface in index.get("prefaceSections", []):
        slug = preface.get("slug", preface.get("file"))
        if slug:
            files_to_dl.append(book_part + "/" + slug + ".json")
    
    # Determine file prefix from book name
    book_name = book_part.split("/")[0]
    if "tefilos" in book_name:
        prefix = "prayer"
    elif "halachos" in book_name:
        prefix = "halacha"
    elif "eitzos" in book_name:
        prefix = "eitzah"
    else:
        prefix = "torah"
    
    # Download torahs - construct filenames from number
    for torah in index.get("torahs", []):
        num = torah.get("number")
        if num:
            files_to_dl.append(book_part + "/" + f"{prefix}-{num}.json")

    count = 0
    with ThreadPoolExecutor(max_workers=50) as executor:
        futures = {executor.submit(dl, f): f for f in files_to_dl}
        for future in as_completed(futures):
            result = future.result()
            if result:
                count += 1
    return count

def main():
    print("=== Downloading reader data ===", flush=True)
    books = [
        "likutay-moharan/part-1",
        "likutay-moharan/part-2",
        "likutay-tefilos/part-1",
        "likutay-tefilos/part-2",
        "likutay-tefilos/part-3",
        "likutay-tefilos/part-4",
        "likutay-halachos/part-1",
        "likutay-halachos/part-2",
        "likutay-halachos/part-3",
        "likutay-halachos/part-4",
        "chumash-lh/part-1",
    ]
    total = 0
    for bp in books:
        print(f"  Downloading {bp}...", flush=True)
        t0 = time.time()
        n = download_book_part(bp)
        elapsed = time.time() - t0
        print(f"    {n} files in {elapsed:.1f}s", flush=True)
        total += n

    try:
        lm_files = os.listdir(os.path.join(DEST, "likutay-moharan/part-1"))
        print(f"  Verification: {len(lm_files)} files in LM part-1", flush=True)
    except:
        print("  Verification failed!", flush=True)

    print(f"Total: {total} files downloaded", flush=True)

if __name__ == "__main__":
    main()

File: scripts/download-reader-data.mjs
/**
 * Download reader JSON data from GitHub raw CDN for Vercel builds.
 * Node.js ESM version.
 */
import https from 'https';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, '..');
const BASE = 'https://raw.githubusercontent.com/petteknanach/ajew-org/main/public/reader';
const DEST = path.join(ROOT, 'public', 'reader');

function dl(filepath) {
  return new Promise((resolve, reject) => {
    const url = BASE + '/' + filepath;
    https.get(url, (res) => {
      if (res.statusCode === 301 || res.statusCode === 302) {
        https.get(res.headers.location, (res2) => {
          if (res2.statusCode !== 200) return resolve(null);
          const data = [];
          res2.on('data', c => data.push(c));
          res2.on('end', () => {
            const dest = path.join(DEST, filepath);
            fs.mkdirSync(path.dirname(dest), { recursive: true });
            fs.writeFileSync(dest, Buffer.concat(data));
            resolve({ file: filepath, size: data.length });
          });
        }).on('error', reject);
        return;
      }
      if (res.statusCode !== 200) return resolve(null);
      const data = [];
      res.on('data', c => data.push(c));
      res.on('end', () => {
        const dest = path.join(DEST, filepath);
        fs.mkdirSync(path.dirname(dest), { recursive: true });
        fs.writeFileSync(dest, Buffer.concat(data));
        resolve({ file: filepath, size: data.length });
      });
    }).on('error', reject);
  });
}

async function getIndex(bookPart) {
  const result = await dl(bookPart + '/index.json');
  if (!result) return null;
  return JSON.parse(fs.readFileSync(path.join(DEST, bookPart, 'index.json'), 'utf8'));
}

async function downloadBookPart(bookPart) {
  const index = await getIndex(bookPart);
  if (!index) return 0;

  const filesToDl = [];
  for (const intro of (index.introSections || [])) {
    const slug = intro.slug || intro.file;
    if (slug) filesToDl.push(bookPart + '/' + slug + '.json');
  }
  for (const preface of (index.prefaceSections || [])) {
    const slug = preface.slug || preface.file;
    if (slug) filesToDl.push(bookPart + '/' + slug + '.json');
  }

  const bookName = bookPart.split('/')[0];
  let prefix = 'torah';
  if (bookName.includes('tefilos')) prefix = 'prayer';
  else if (bookName.includes('halachos')) prefix = 'halacha';
  else if (bookName.includes('eitzos')) prefix = 'eitzah';

  for (const torah of (index.torahs || [])) {
    const num = torah.number;
    if (num) filesToDl.push(bookPart + '/' + prefix + '-' + num + '.json');
  }

  let count = 0;
  // Process in batches of 50 for parallel downloads
  for (let i = 0; i < filesToDl.length; i += 50) {
    const batch = filesToDl.slice(i, i + 50);
    const results = await Promise.allSettled(batch.map(f => dl(f)));
    count += results.filter(r => r.status === 'fulfilled' && r.value).length;
  }
  return count;
}

async function main() {
  console.log('=== Downloading reader data ===');
  const books = [
    'likutay-moharan/part-1',
    'likutay-moharan/part-2',
    'likutay-tefilos/part-1',
    'likutay-tefilos/part-2',
    'likutay-halachos/part-1',
    'chumash-lh/part-1',
  ];

  let total = 0;
  for (const bp of books) {
    console.log('  Downloading ' + bp + '...');
    const t0 = Date.now();
    const n = await downloadBookPart(bp);
    const elapsed = ((Date.now() - t0) / 1000).toFixed(1);
    console.log('    ' + n + ' files in ' + elapsed + 's');
    total += n;
  }

  try {
    const lmFiles = fs.readdirSync(path.join(DEST, 'likutay-moharan/part-1'));
    console.log('  Verification: ' + lmFiles.length + ' files in LM part-1');
  } catch (e) {
    console.log('  Verification failed: ' + e.message);
  }
  console.log('Total: ' + total + ' files downloaded');
}

main().catch(e => {
  console.error('Failed:', e.message);
  process.exit(1);
});
